File: src/main/java/com/fwdekker/randomness/ui/JBPopupHelper.java
package com.fwdekker.randomness.ui;

import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.ui.popup.list.ListPopupImpl;
import com.intellij.ui.speedSearch.SpeedSearch;

import javax.swing.AbstractAction;
import javax.swing.KeyStroke;
import java.awt.event.ActionEvent;
import java.awt.event.KeyEvent;
import java.util.Locale;
import java.util.stream.IntStream;


/**
 * A collection of helper methods for dealing with {@link com.intellij.openapi.ui.popup.JBPopup}s.
 */
public final class JBPopupHelper {
    private static final int NINE = 9;


    /**
     * Private constructor to prevent instantiation.
     */
    private JBPopupHelper() {
        // Do nothing
    }


    /**
     * Disables speed search.
     *
     * @param popup the popup to disable speed search for
     */
    public static void disableSpeedSearch(final ListPopupImpl popup) {
        final SpeedSearch speedSearch = popup.getSpeedSearch();
        speedSearch.setEnabled(false);
        speedSearch.addChangeListener(event -> speedSearch.updatePattern(""));
    }

    /**
     * Registers actions such that actions can be selected while holding {@code Shift}.
     *
     * @param popup       the popup to enable selecting with {@code Shift} for
     * @param normalTitle the title of the popup while the {@code Shift} key is not pressed
     * @param shiftTitle  the title of the popup while the {@code Shift} key is pressed
     */
    public static void registerShiftActions(final ListPopupImpl popup, final String normalTitle,
                                            final String shiftTitle) {
        registerModifierActions(popup, ModifierKey.SHIFT, normalTitle, shiftTitle);
    }

    /**
     * Registers actions such that actions can be selected while holding {@code Ctrl} or {@code Cmd}.
     *
     * @param popup       the popup to enable selecting with {@code Ctrl}/{@code Cmd} for
     * @param normalTitle the title of the popup while the {@code Ctrl}/{@code Cmd} key is not pressed
     * @param ctrlTitle   the title of the popup while the {@code Ctrl}/{@code Cmd} key is pressed
     */
    public static void registerCtrlActions(final ListPopupImpl popup, final String normalTitle,
                                           final String ctrlTitle) {
        registerModifierActions(popup, ModifierKey.CTRL, normalTitle, ctrlTitle);
    }

    /**
     * Registers actions such that actions can be selected while holding {@code Ctrl} or {@code Cmd}.
     *
     * @param popup         the popup to enable selecting with the modifier key for
     * @param modifierKey   the modifier key for which actions should be registered
     * @param normalTitle   the title of the popup while the modifier key is not pressed
     * @param modifierTitle the title of the popup while the modifier key is pressed
     */
    private static void registerModifierActions(final ListPopupImpl popup, final ModifierKey modifierKey,
                                                final String normalTitle, final String modifierTitle) {
        final String lcShortModifierName = modifierKey.shortName.toLowerCase(Locale.getDefault());
        final String lcLongModifierName = modifierKey.longName.toLowerCase(Locale.getDefault());
        final String ucLongModifierName = modifierKey.longName.toUpperCase(Locale.getDefault());

        popup.registerAction(
            lcShortModifierName + "Released",
            KeyStroke.getKeyStroke("released " + ucLongModifierName),
            new AbstractAction() {
                @Override
                public void actionPerformed(final ActionEvent event) {
                    popup.setCaption(normalTitle);
                }
            }
        );
        popup.registerAction(
            lcShortModifierName + "Pressed",
            KeyStroke.getKeyStroke(lcLongModifierName + " pressed " + ucLongModifierName),
            new AbstractAction() {
                @Override
                public void actionPerformed(final ActionEvent event) {
                    popup.setCaption(modifierTitle);
                }
            }
        );
        popup.registerAction(
            lcShortModifierName + "InvokeAction",
            KeyStroke.getKeyStroke(lcLongModifierName + " ENTER"),
            new AbstractAction() {
                @Override
                public void actionPerformed(final ActionEvent event) {
                    final KeyEvent keyEvent = new KeyEvent(
                        popup.getComponent(), event.getID(), event.getWhen(),
                        event.getModifiers(), KeyEvent.VK_ENTER, KeyEvent.CHAR_UNDEFINED,
                        KeyEvent.KEY_LOCATION_UNKNOWN
                    );
                    popup.handleSelect(true, keyEvent);
                }
            }
        );

        IntStream.range(1, NINE).forEach(key -> popup
            .registerAction(
                lcShortModifierName + "Invoke" + key,
                KeyStroke.getKeyStroke(lcLongModifierName + " " + key),
                new AbstractAction() {
                    @Override
                    public void actionPerformed(final ActionEvent event) {
                        final KeyEvent keyEvent = new KeyEvent(
                            popup.getComponent(), event.getID(),
                            event.getWhen(), event.getModifiers(),
                            KeyEvent.VK_ENTER,
                            KeyEvent.CHAR_UNDEFINED,
                            KeyEvent.KEY_LOCATION_UNKNOWN
                        );
                        popup.getList().addSelectionInterval(key - 1, key - 1);
                        popup.handleSelect(true, keyEvent);
                    }
                }
            )
        );
    }

    /**
     * Displays a popup with a title and two messages.
     *
     * @param title    the title of the popup
     * @param messageA the first message
     * @param messageB the second message
     */
    public static void showMessagePopup(final String title, final String messageA, final String messageB) {
        final Runnable nop = () -> {
        };

        JBPopupFactory.getInstance().createConfirmation(
            title,
            messageA,
            messageB,
            nop,
            1
        ).showInFocusCenter();
    }


    /**
     * Pairs the short and long name of a modifier key together.
     */
    private enum ModifierKey {
        ALT("alt", "alt"),
        CTRL("ctrl", "control"),
        SHIFT("shift", "shift");


        private final String shortName;
        private final String longName;


        /**
         * Constructs a new modifier key.
         *
         * @param shortName the short name of the modifier key
         * @param longName  the long name of the modifier key
         */
        ModifierKey(final String shortName, final String longName) {
            this.shortName = shortName;
            this.longName = longName;
        }
    }
}

File: src/main/kotlin/com/fwdekker/randomness/ui/JBPopupHelper.kt
package com.fwdekker.randomness.ui

import com.intellij.ui.popup.list.ListPopupImpl
import java.awt.event.ActionEvent
import java.awt.event.KeyEvent
import java.util.Locale
import javax.swing.AbstractAction
import javax.swing.KeyStroke


/**
 * Registers actions such that actions can be selected while holding (combinations of) modifier keys.
 *
 * All combinations of modifier keys are registered for events. Additionally, the [captionModifier] function is invoked
 * every time the user presses or releases any modifier key, even while holding other modifier keys.
 *
 * Events are also registered for pressing the Enter key (with or without modifier keys) to invoke the action that is
 * currently highlighted, and for pressing one of the numbers 1-9 (with or without modifier keys) to invoke the action
 * at that index in the popup.
 *
 * @param captionModifier returns the caption to set based on the event
 */
fun ListPopupImpl.registerModifierActions(captionModifier: (ActionEvent?) -> String) {
    val modifiers = listOf(listOf("alt"), listOf("control"), listOf("shift"))
    val optionalModifiers = listOf(listOf("")) + modifiers

    (modifiers * optionalModifiers * optionalModifiers).forEach { (a, b, c) ->
        registerAction(
            "${a}${b}${c}Released",
            KeyStroke.getKeyStroke("$b $c released ${a.uppercase(Locale.getDefault())}"),
            actionListener { setCaption(captionModifier(it)) }
        )
        registerAction(
            "${a}${b}${c}Pressed",
            KeyStroke.getKeyStroke("$a $b $c pressed ${a.uppercase(Locale.getDefault())}"),
            actionListener { setCaption(captionModifier(it)) }
        )
        registerAction(
            "${a}${b}${c}invokeAction",
            KeyStroke.getKeyStroke("$a $b $c pressed ENTER"),
            actionListener { event ->
                event ?: return@actionListener

                handleSelect(
                    true,
                    KeyEvent(
                        component,
                        event.id, event.getWhen(), event.modifiers,
                        KeyEvent.VK_ENTER, KeyEvent.CHAR_UNDEFINED, KeyEvent.KEY_LOCATION_UNKNOWN
                    )
                )
            }
        )

        @Suppress("MagicNumber") // Not worth a constant
        for (key in 0..9) {
            registerAction(
                "${a}${b}${c}invokeAction$key",
                KeyStroke.getKeyStroke("$a $b $c pressed $key"),
                actionListener { event ->
                    event ?: return@actionListener

                    val targetRow = if (key == 0) 9 else key - 1
                    list.addSelectionInterval(targetRow, targetRow)
                    handleSelect(
                        true,
                        KeyEvent(
                            component,
                            event.id, event.getWhen(), event.modifiers,
                            KeyEvent.VK_ENTER, KeyEvent.CHAR_UNDEFINED, KeyEvent.KEY_LOCATION_UNKNOWN
                        )
                    )
                }
            )
        }
    }
}

/**
 * Returns an `AbstractAction` that uses [actionPerformed] as the implementation of its `actionPerformed` method.
 *
 * @param actionPerformed the code to execute in `actionPerformed`
 * @return an `AbstractAction` that uses [actionPerformed] as the implementation of its `actionPerformed` method
 */
private fun actionListener(actionPerformed: (ActionEvent?) -> Unit) =
    object : AbstractAction() {
        override fun actionPerformed(event: ActionEvent?) = actionPerformed(event)
    }

/**
 * Returns the cartesian product of two lists.
 *
 * By requiring both lists to actually be lists of lists, this method can be chained.
 *
 * Consider the following examples, using a simplified notation for lists for readability:
 * ```
 * $ [[1, 2]] * [[3, 4]]
 * [[1, 3], [1, 4], [2, 3], [2, 4]]
 *
 * $ [[1, 2]] * [[3, 4]] * [[5, 6]]
 * [[1, 3, 5], [1, 3, 6], [1, 4, 5], [1, 4, 6], [2, 3, 5], [2, 3, 6], [2, 4, 5], [2, 4, 6]]
 * ```
 *
 * @param E the type of inner element
 * @param other the list to multiply with
 * @return the cartesian product of `this` and [other]
 */
@Suppress("UnusedPrivateMember") // False positive: Used as operator `*`
private operator fun <E> List<List<E>>.times(other: List<List<E>>) =
    this.flatMap { t1 -> other.map { t2 -> t1 + t2 } }
